state/storage.go

```diff
@@ -120,17 +120,37 @@ func (s *Storage) InsertAccountData(userID, roomID string, events []json.RawMess
 	return data, err
 }
 
+// Prepare a snapshot of the database for calling snapshot functions.
+func (s *Storage) PrepareSnapshot(txn *sqlx.Tx) (tableName string, err error) {
+	// create a temporary table with all the membership nids for the current snapshots for all rooms.
+	// A temporary table will be deleted when the postgres session ends (this process quits).
+	// We insert these into a temporary table to let the query planner make better decisions. In practice,
+	// if we instead nest this SELECT as a subselect, we see very poor query times for large tables as
+	// each event NID is queried using a btree index, rather than doing a seq scan as this query will pull
+	// out ~50% of the rows in syncv3_events.
+	tempTableName := "temp_snapshot"
+	_, err = txn.Exec(
+		`SELECT UNNEST(membership_events) AS membership_nid INTO TEMP ` + tempTableName + ` FROM syncv3_snapshots
+		JOIN syncv3_rooms ON syncv3_snapshots.snapshot_id = syncv3_rooms.current_snapshot_id`,
+	)
+	return tempTableName, err
+}
+
```

Comment:
Presumably if you call this twice then you'll end up duplicating all rows in the temp table?

Reply:
(Not that we ever do call it twice, of course)

Reply:
It would return an error actually as the table will already exist.